src/OrbitGl/TrackContainer.cpp

```diff
@@ -370,11 +369,63 @@ void TrackContainer::SetIteratorOverlayData(
   RequestUpdate(RequestUpdateScope::kDraw);
 }
 
+void TrackContainer::DrawThreadDependencyArrow(PrimitiveAssembler& primitive_assembler,
+                                               PickingMode picking_mode) {
+  std::optional<orbit_client_data::ThreadStateSliceInfo> thread_state_slice =
+      app_->selected_thread_state_slice();
+  if (!thread_state_slice.has_value() || picking_mode != PickingMode::kNone) {
+    return;
+  }
+
+  orbit_client_data::ThreadID start_arrow_thread = thread_state_slice->wakeup_tid();
+  orbit_client_data::ThreadID end_arrow_thread = thread_state_slice->tid();
+
+  std::optional<ThreadTrack*> start_arrow_track_opt =
+      track_manager_->GetThreadTrack(start_arrow_thread);
+  std::optional<ThreadTrack*> end_arrow_track_opt =
+      track_manager_->GetThreadTrack(end_arrow_thread);
+
+  if (!start_arrow_track_opt.has_value() || !end_arrow_track_opt.has_value()) {
+    return;
+  }
+
+  ThreadTrack* start_arrow_track = start_arrow_track_opt.value();
+  ThreadTrack* end_arrow_track = end_arrow_track_opt.value();
+
+  const std::vector<Track*>& threads = track_manager_->GetVisibleTracks();
+
+  if (std::find(threads.begin(), threads.end(), start_arrow_track) == threads.end() ||
+      std::find(threads.begin(), threads.end(), end_arrow_track) == threads.end()) {
+    return;
+  }
+
+  const Color kArrowColor(255, 255, 255, 255);
+  const float start_arrow_thread_state_half_height =
+      start_arrow_track->GetThreadStateBarHeight() / 2;
+  const float end_arrow_thread_state_half_height = end_arrow_track->GetThreadStateBarHeight() / 2;
+  float x = timeline_info_->GetWorldFromTick(thread_state_slice->begin_timestamp_ns());
+  float start_arrow_y =
+      start_arrow_track->GetThreadStateBarPos()[1] + start_arrow_thread_state_half_height;
+  float end_arrow_y =
+      end_arrow_track->GetThreadStateBarPos()[1] + end_arrow_thread_state_half_height;
+  float arrow_total_height = std::abs(end_arrow_y - start_arrow_y);
+  float body_height = arrow_total_height - layout_->GetThreadDependencyArrowHeadHeight();
+  PrimitiveAssembler::ArrowDirection arrow_direction =
+      start_arrow_y < end_arrow_y ? PrimitiveAssembler::ArrowDirection::kDown
+                                  : PrimitiveAssembler::ArrowDirection::kUp;
+  primitive_assembler.AddVerticalArrow(
+      Vec2{x, start_arrow_y}, Vec2{layout_->GetThreadDependencyArrowBodyWidth(), body_height},
+      Vec2{layout_->GetThreadDependencyArrowHeadWidth(),
+           layout_->GetThreadDependencyArrowHeadHeight()},
+      IntrospectionWindow::kZValueOverlay, kArrowColor, arrow_direction);
```

Comment:
1) `GlCanvas::kZValueOverlay`, you shouldn't use InstrospectionWindow here as it's completely unrelated.

And just a comment: In general Overlays are used to transparent layers that are drawn on top of everything, but in this case is something that will appear after clicking, so I tend to think that `kZValueUi` might be better. Anyway, I would like to manually test it and later we can change it if we prefer.